packages/eslint-plugin/tests/rules/no-array-delete.test.ts

```diff
@@ -0,0 +1,390 @@
+import rule from '../../src/rules/no-array-delete';
+import { getFixturesRootDir, RuleTester } from '../RuleTester';
+
+const rootDir = getFixturesRootDir();
+const ruleTester = new RuleTester({
+  parserOptions: {
+    ecmaVersion: 2015,
+    tsconfigRootDir: rootDir,
+    project: './tsconfig.json',
+  },
+  parser: '@typescript-eslint/parser',
+});
+
+ruleTester.run('no-array-delete', rule, {
+  valid: [
+    `
+declare const obj: Record<PropertyKey, unknown>;
+declare const key: PropertyKey;
+delete obj[key];
+    `,
+    `
+declare const arr: unknown[];
+delete arr.myprop;
+    `,
+    `
+declare const arr: unknown[];
+declare const i: string;
+delete arr[i];
+    `,
+    `
+declare const multiDimesnional: Array<unknown>[][][][];
+declare const i: number;
+delete multiDimesnional[i][i][i][i][i].someProp;
+    `,
+  ],
+  invalid: [
```

Comment:
A few more test cases off the top of my head:

* `ReadonlyArray<...>`
* `readonly ...[]`
* `Readonly<...[]>`
* `Readonly<Array<...>>`